### ISSUE TITLE
Specialize FMA Accumulators Properly

### ISSUE THREAD

**Issue by mattfel1** [stanford-ppl/spatial#63]
If we didn't turn `FixAdd(FixMul(a,b),c)` into `FixFMA(a,b,c)`, then a `Reduce(0)(...){i => i*i}{_+_}` would not have a huge II, since the FixMul would be outside of the reduction loop between RegRead, FixAdd, RegWrite.  But assuming we are to keep this transform, I am working on a specialized accumulator for this case that has as many lanes as II and fuses the lanes in a reduction tree that allows the effective II = 1.

**Comment by mattfel1** [comment]
At unrolling time, we see a FixAdd as the `reduce.result` node and set everybody in the cycle as FixPtSum.  But then at RewriteTransformer time, we sometimes change this add to `FixFMA`.  It's easy to update the reduceType on this node to be `FixFMA`, but not sure what the best way to propogate this info to everyone else in the cycle.  For now I just have it being done in the `@stateful def latencyAndInterval` but I think the better solution is to probably have every node in the reduction cycle stored as metadata, and then make a flow that propogates the reduceType and gives preference to FixFMA over FixAdd?

**Comment by mattfel1** [comment]
Not at all proud of the changes I just made... But at least I have it working so that we can detect FixPtSum accumulations from unrolling that change into FixPtFMA accumulations by setting the metadata in the RewriteTransformer.

Not sure if we want to have a specialized reduction node that takes only the new value(s) (mul1 mul2 for fma, just one element for sum, etc), or if we want the full loop in the IR and want codegen to do something special.  Right now, I made it so that info about an FMA accumulation cycle is spread around all of the nodes that are part of the cycle (using the retime transformer).  It takes the data from the RegWrite and updates all its dependencies' latency paths by the cycles log2(ii). It forces the II of this cycle to 1.  Then in codegen I ignore emission for nodes part of an fma accumulation and drop in the specialized template and wire it in.

I wanted to make sure the templates and everything are in order, which seems to be the case, so now stefan can have all the optimizations he wanted.  We should sit down and figure out exactly how to integrate this stuff properly though.
